docs/dev-guides/java-style.md

```diff
@@ -0,0 +1,166 @@
+
+_Exception_: Stairway is not Spring-aware, so it is a common pattern to create a dependency-injected class that wraps Stairway functionality.
+
+### Exceptions
+
+**Checked exceptions** place a burden on the users of an API. They are a tool best used when _both_ of the following are true: (1) the exception cannot be prevented by proper use of the API, and (2) the user can take some meaningful action when encountering the exception. Unless both conditions are met, checked exceptions _should_ be avoided. (See [EJ3 #71](https://github.com/david-sauvage/effective-java-summary#exceptions) for more discussion.)
+
+Guidance for **unchecked exceptions**:
+
+*   More specific exceptions _should_ be thrown instead of generic RuntimeExceptions.
+    *   IllegalArgumentException, ConcurrentModificationException, UnsupportedOperationException are commonly reused exceptions from the Java library.
+    *   Other Terra-specific runtime exceptions _should_ be centralized in [terra-common-lib](https://github.com/DataBiosphere/terra-common-lib/tree/develop/src/main/java/bio/terra/common/exception).
+*   Service repos may also house application-specific exceptions ([WSM example](https://github.com/DataBiosphere/terra-workspace-manager/tree/dev/src/main/java/bio/terra/workspace/common/exception)).
+*   In many cases, exceptions are explicitly associated with an HTTP status code ([example](https://github.com/DataBiosphere/terra-workspace-manager/blob/4a33150f2143d163e89bfdda839e4ebf04f09b03/src/main/java/bio/terra/workspace/common/exception/ErrorReportException.java#L11)) and used to auto-generate an appropriate response ([example](https://github.com/DataBiosphere/terra-workspace-manager/blob/4a33150f2143d163e89bfdda839e4ebf04f09b03/src/main/java/bio/terra/workspace/app/controller/GlobalExceptionHandler.java#L33)).
+
+### Final
+
+*   **Classes** – unless a class has been designed and documented for inheritance, it _should_ be marked as final.
+*   **Function parameters **– _prefer_ not marking as final, due to the noisiness of the “final” keyword. Use where necessary for occasional semantic emphasis.
+*   **Instance variables** – _always_ mark as final if it is never reassigned.
+*   **Local variables** – use of final is _optional_. Short and simple methods can help avoid inappropriate variable reassignment.
+*   **Static variables** – almost _always_ mark as final.
+
+### Immutability
+
+Quoting from Effective Java: "immutable classes are easier to design, implement, and use than mutable classes. They are less prone to error and are more secure." See [recommendations from Effective Java](https://github.com/HugoMatilla/Effective-JAVA-Summary#15-minimize-mutability) on how to minimize mutability. [Immutables](https://immutables.github.io/) is a nice library for immutable data transfer objects, as is [Guava](https://github.com/google/guava/wiki/ImmutableCollectionsExplained) for immutable collections.
```

Comment:
Are we recommending Immutables and Guava? I notice this doesn't use any of the always, should, prefer, may vocabulary.

Reply:
I tried to re-phrase a bit. There's a delicate balance here, between this guide becoming a laundry list of libraries used throughout the repo, and ignoring important best practices. I'm leaning towards calling out libraries that have a sizable impact on a key code aesthetic, and where multiple options may exist. I updated the language to also call out that Guava and Autovalue are in active use, while Immutables is under consideration.